xdsl/backend/riscv/lowering/utils.py

```diff
@@ -0,0 +1,73 @@
+from xdsl.dialects import builtin, riscv
+from xdsl.ir import Block, SSAValue
+from xdsl.pattern_rewriter import PatternRewriter
+
+
+def cast_operands_to_int_regs(rewriter: PatternRewriter) -> list[SSAValue]:
+    """
+    Add cast operations just before the targeted operation
+    if the operands were not already int registers
+    """
+
+    # convert all values to int registers if they are not already
+    casted_ops = [
+        builtin.UnrealizedConversionCastOp.get(
+            [not_reg], (riscv.IntRegisterType.unallocated(),)
+        )
+        for not_reg in filter(
+            lambda v: not isinstance(v.type, riscv.IntRegisterType),
+            rewriter.current_operation.operands,
+        )
+    ]
+
+    rewriter.insert_op_before_matched_op(casted_ops)
+
+    # return the original value if it's already an int register
+    # otherwise return the casted value
+    return [
+        casted_ops.pop(0).results[0]
+        if not isinstance(operand.type, riscv.IntRegisterType)
+        else operand
+        for operand in rewriter.current_operation.operands
+    ]
```

Comment:
```suggestion
    new_ops = list[Operation]()
    new_operands = list[SSAValue]()
    
    for operand in rewriter.current_operation.operands:
        if not isinstance(v.type, riscv.IntRegisterType):
            cast_op = builtin.UnrealizedConversionCastOp.get(
                (not_reg,), (riscv.IntRegisterType.unallocated(),)
            )
            new_ops.append(cast_op)
            operand = cast_op.results[0]
        new_operands.append(operand)
    
    rewriter.insert_op_before_matched_op(new_ops)
    return new_operands
```

I find that list comprehensions that mutate are a bit harder to understand than for loops, in this case the idea can be expressed a bit more explicitly in a loop, I think. There's also a way to do this with no for loops by tracking indices but I'm not sure it's worth it.